drivers/serial/pty.c

```diff
@@ -583,31 +516,23 @@ static ssize_t pty_read(FAR struct file *filep, FAR char *buffer, size_t len)
 
           if (ch != '\r' || (dev->pd_iflag & IGNCR) == 0)
             {
-              /* Transfer the (possibly translated) character and update the
-               * count of bytes transferred.
-               */
-
-              *buffer++ = ch;
-              ntotal++;
+              buffer[j++] = ch;
             }
         }
-    }
-  else
-#endif
-    {
-      /* NOTE: the source pipe will block if no data is available in
-       * the pipe.   Otherwise, it will return data from the pipe.  If
-       * there are fewer than 'len' bytes in the, it will return with
-       * ntotal < len.
-       *
-       * REVISIT: Should not block if the oflags include O_NONBLOCK.
-       * How would we ripple the O_NONBLOCK characteristic to the
-       * contained source pipe? file_fcntl()?  Or FIONREAD?  See the
-       * TODO comment at the top of this file.
-       */
 
-      ntotal = file_read(&dev->pd_src, buffer, len);
+      /* Is the buffer empty after process? */
+
+      if (j == 0)
+        {
+          /* Yes, try again */
+
+          ntotal = file_read(&dev->pd_src, buffer, len);
+          goto again;
```

Comment:
Instead using goto and again label, why not do this way:
```
while (ntotal = file_read(&dev->pd_src, buffer, len) != 0)
{
  for (i = j = 0; i < ntotal; i++)
    {
      ...
      if (j == 0)
      {
         break;
      }
    }
}
```

Reply:
do you mean: `while (ntotal = file_read(&dev->pd_src, buffer, len) > 0)`?

Reply:
I think should be:
```
    if (j == 0)
      {
        break;
      }
```
This is like @xiaoxiang781216 did in the original code. Before doing "goto again" he did ntotal=file_read(), but we don't need to do it, because after the break it will be done in the while().

Reply:
> Instead using goto and again label, why not do this way:
> 
> ```
> while (ntotal = file_read(&dev->pd_src, buffer, len) != 0)
> {
>   for (i = j = 0; i < ntotal; i++)
>     {
>       ...
>       if (j == 0)
>       {
>          break;
>       }
>     }
> }
> ```

Because the first read is at:
https://github.com/apache/incubator-nuttx/pull/5569/files#diff-16e2adad75350dfb43ccb718a904622b997cff755e6de9953e44c654ef48661eR476
we have to skip it.

Reply:
@acassis is proposing to rework the first read as well and put it into a `while()`

Reply:
Like
```
while (ntotal = file_read(&dev->pd_src, buffer, len) > 0)
  {
#ifdef CONFIG_SERIAL_TERMIOS
    if (dev->pd_iflag & (INLCR | IGNCR | ICRNL))
      {
        for (i = j = 0; i < ntotal; i++)
          {
          }
        
        if (j != 0)
          {
            ntotal = j;
            break;
          }
      }
#endif
  }

return ntotal;
```